libs/metrics/src/tokio_metrics.rs

```diff
@@ -0,0 +1,263 @@
+    ) -> proto::MetricFamily {
+        self.metric(desc, proto::MetricType::GAUGE, |rt| {
+            let mut g = proto::Gauge::default();
+            g.set_value(f(rt));
+            let mut m = proto::Metric::default();
+            m.set_gauge(g);
+            m
+        })
+    }
+
+    // fn guage_per_worker(
+    //     &self,
+    //     desc: &Desc,
+    //     f: impl Fn(&tokio::runtime::RuntimeMetrics, usize) -> f64,
+    // ) -> proto::MetricFamily {
+    //     self.metric_per_worker(desc, proto::MetricType::GAUGE, |m, i| {
+    //         let mut g = proto::Gauge::default();
+    //         g.set_value(f(m, i));
+    //         let mut m = proto::Metric::default();
+    //         m.set_gauge(g);
+    //         m
+    //     })
+    // }
+
+    fn counter_per_worker(
+        &self,
+        desc: &Desc,
+        f: impl Fn(&tokio::runtime::RuntimeMetrics, usize) -> f64,
+    ) -> proto::MetricFamily {
+        self.metric_per_worker(desc, proto::MetricType::COUNTER, |m, i| {
+            let mut g = proto::Counter::default();
+            g.set_value(f(m, i));
+            let mut m = proto::Metric::default();
+            m.set_counter(g);
+            m
+        })
+    }
+
+    fn metric(
+        &self,
+        desc: &Desc,
+        typ: proto::MetricType,
+        f: impl Fn(&tokio::runtime::RuntimeMetrics) -> proto::Metric,
+    ) -> proto::MetricFamily {
+        self.metrics(desc, typ, |rt, label_id, metrics| {
+            let mut m = f(rt);
+            m.set_label(vec![label_id.clone()]);
+            metrics.push(m);
+        })
+    }
+
+    fn metric_per_worker(
+        &self,
+        desc: &Desc,
+        typ: proto::MetricType,
+        f: impl Fn(&tokio::runtime::RuntimeMetrics, usize) -> proto::Metric,
+    ) -> proto::MetricFamily {
+        self.metrics(desc, typ, |rt, label_id, metrics| {
+            let workers = rt.num_workers();
+            for i in 0..workers {
+                let mut m = f(rt, i);
+
+                let mut label_worker = LabelPair::default();
+                label_worker.set_name("worker".to_owned());
+                label_worker.set_value(i.to_string());
+
+                m.set_label(vec![label_id.clone(), label_worker]);
+                metrics.push(m);
+            }
+        })
+    }
+
+    fn metrics(
+        &self,
+        desc: &Desc,
+        typ: proto::MetricType,
+        f: impl Fn(&tokio::runtime::RuntimeMetrics, LabelPair, &mut Vec<proto::Metric>),
+    ) -> proto::MetricFamily {
+        let mut m = proto::MetricFamily::default();
+        m.set_name(desc.fq_name.clone());
+        m.set_help(desc.help.clone());
+        m.set_field_type(typ);
+        let mut metrics = vec![];
+        for rt in &*self.handles.lock().unwrap() {
+            let rt_metrics = rt.metrics();
+
+            let mut label_id = LabelPair::default();
+            label_id.set_name("id".to_owned());
+            label_id.set_value(rt.id().to_string());
```

Comment:
Unsure if it makes sense to use runtime id as there are no guarantees with it, and we would have to then handle this id => name mapping elsewhere; id's could theoretically be different across starts etc..

Perhaps a design instead where there is a single immutable `HashMap<&'static str, Handle>` instead of the mutable vec of handles?

Reply:
I think we can keep the ID - it's at least unique per pod which our prometheus metrics are separated by. But a runtime name would also make sense to have associated.

Reply:
They are here: https://github.com/neondatabase/neon/blob/ba4fe9e10f62dea18e5aa0f26b43a60195397818/pageserver/src/task_mgr.rs#L106-L150 -- all static. It would be quite maddening if they were dynamic :)

Need to make sure safekeepers are similar.